[Feature] CreateView project #78

Gojaehyeon opened the pull request:
## 관련 이슈
- Closes #78
- Closes #80

## 작업 내용
- 프로젝트 생성시 SwiftData와 연결하여 저장 구현
- 프로젝트 신규 생성시 기존 프로젝트 완료되도록 구현(로직은 기존에도 존재)
- 프로젝트 생성 모달에서 프로젝트 이름 우측에 진행 날짜 표시
- 사진 추가시 더미이미지가 아닌 실제 이미지와 연결되도록 구현
- 모든 이미지 최신순으로 배열되도록 수정
- 작성 완료 탭시 dismiss되도록 구현
- 작성중인 항목 존재할 경우 X버튼 눌러도 재확인 절차 추가

## 스크린샷 (UI 변경 시)
![ScreenRecording_06-06-2025 19-54-18_1](https://github.com/user-attachments/assets/1ca25dcf-2514-4320-aa07-1973cd61bfee)


## 체크리스트
- [x] 빌드 에러 없음
- [ ] 기본 동작 테스트 완료
- [ ] 코드 리뷰 준비 완료

Gojaehyeon committed:
feat: Document 기반 이미지 저장/로드, 정렬 로직 개선, 정사각형 이미지 UI 적용

- 이미지 저장 및 로드 방식을 Document 디렉토리 기반으로 수정
- 모든 포스트 최신순 정렬 적용 (ProjectSectionCard, ActiveProjectCard, MainView 그리드)
- GridImageCard에 GeometryReader 적용해 정사각형 이미지 표시
- 이미지가 없는 경우 메모 텍스트 표시 개선
- 이미지 로드 실패 시 placeholder 이미지 추가
- Post 모델의 이미지 URL 저장 방식 개선 및 SwiftData 연동 최적화

Gojaehyeon committed:
feat: 작성 완료 누를 시 createview dismiss 되는 기능 추가

Gojaehyeon committed:
feat: 과정 기록중 x 누를 경우 재확인 절차 추가

bishoe01 commented on C3-4T2D/View/Create/CreateView.swift:
```diff
@@ -102,6 +117,14 @@ struct CreateView: View {
         .onDisappear {
             createPickedImage = nil
         }
+        .alert("작성 중인 내용이 있어요", isPresented: $showExitAlert) {
+            Button("취소", role: .cancel) { }
+            Button("종료", role: .destructive) {
+                dismiss()
+            }
+        } message: {
+            Text("정말 종료하시겠어요?")
+        }
```
중간 이탈에 대해 처리 좋네요 👍

bishoe01 commented on C3-4T2D/View/Create/Local/ProjectSelector.swift:
```diff
@@ -101,21 +115,29 @@ struct ProjectSelector: View {
         guard isAddingProject, !trimmedName.isEmpty else { return }
 
-        projects.append(trimmedName)
+        // 기존 진행중인 프로젝트 완료 처리
+        if let current = projects.first(where: { $0.finishedAt == nil }) {
+            current.finishedAt = Date()
+        }
+        let newProject = Project(projectTitle: trimmedName)
+        context.insert(newProject)
+        try? context.save()
```
이부분 `SwiftManager`에 있는 `startNewProject` 랑 같은 로직인데,  title에 `trimmedName` 넣어서 다듬어줌녀 될 것 같은데 가능하실까요 ? 

```swift
/// 새 프로젝트 시작 ( 진행중인 프로젝트가 있으면 바로 완료 처리! )
    static func startNewProject(context: ModelContext, title: String) -> Project {
        // 기존 진행중인 프로젝트 완료 처리
        if let currentProject = getCurrentProject(context: context) {
            // 바로 finished에 날짜 집어넣기 -> 이러면 완료처리임 `!= nil`
            currentProject.finishedAt = Date()
            print("현 \(currentProject.projectTitle)")
        }

        // 새로운 프로젝트 생성 ,title만 먼저 받음
        let newProject = Project(projectTitle: title)
        context.insert(newProject)

        do {
            try context.save()
            print("새로 만든 프로젝트 이름 : \(title)")
        } catch {
            print("프로젝트 새로만들기 에러 \(error)")
        }

        return newProject
    }

```

Gojaehyeon replied:
넵!! 다듬어서 커밋 올릴게요!

Gojaehyeon replied:
이부분 커밋 올렸어용

bishoe01 commented on C3-4T2D/View/Create/Local/ProjectSelector.swift:
```diff
@@ -101,21 +115,29 @@ struct ProjectSelector: View {
+        formatter.dateFormat = "yyyy.MM.dd"
+        return formatter.string(from: date)
     }
 }
```
이 부분 `DateExtensions.swift`에 있는 projectDateRange 초반 부분과 일치하는데, 이 부분을 따로 분리해서 여기 파일에서 `dateString` 대신활용하는건 어떨까요? 
```swift
static func projectDateRange(startDate: Date, endDate: Date?) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy.MM.dd"

        if let endDate = endDate {
            return "\(formatter.string(from: startDate)) ~ \(formatter.string(from: endDate))"
        } else {
            return "\(formatter.string(from: startDate)) ~"
        }
    }
    ```

Gojaehyeon replied:
오 그러네요! 이부분도 수정해서 커밋할게용

Gojaehyeon replied:
이부분도 커밋 완료

bishoe01 reviewed (changes requested):
이미지 관련 기능이 추가됐네요 고생하셨습니다! 
말씀드린 부분만 존재하는 부분에서 활용해주시면 좋을 것 같습니다 !

chawj1234 commented on C3-4T2D/View/Create/CreateView.swift:
```diff
@@ -75,9 +89,10 @@ struct CreateView: View {
+                            .background((selectedProject != nil && (!descriptionText.isEmpty || createPickedImage != nil)) ? Color.prime1 : Color.gray)
                             .cornerRadius(8)
                     }
+                    .disabled(selectedProject == nil || (descriptionText.isEmpty && createPickedImage == nil))
```
'작성 완료' 버튼 활성화 로직을 깔끔하게 구현하셨네요👍

chawj1234 commented on C3-4T2D/View/Create/Local/ProjectSelector.swift:
```diff
@@ -91,7 +104,8 @@ struct ProjectSelector: View {
         .confirmationDialog("정말 삭제하시겠습니까?", isPresented: $showDeleteConfirmation, titleVisibility: .visible) {
             Button("삭제", role: .destructive) {
                 if let project = projectToDelete {
-                    deleteProject(project)
+                    context.delete(project)
+                    try? context.save()
                 }
                 projectToDelete = nil
             }
```
선택된 프로젝트를 삭제하려고 할 때 앱이 크래시나는 문제가 있습니다. 선택된 프로젝트를 삭제하려는 경우를 고려한 예외 처리가 필요해 보입니다!

chawj1234 replied:
다시 확인해보니 선택 유무에 상관없이 크래시가 발생하네요

Gojaehyeon replied:
조치해서 커밋 올릴게요~

Gojaehyeon replied:
조치완료!

chawj1234 commented on C3-4T2D/View/Local/PostListCard.swift:
```diff
@@ -10,13 +10,43 @@ import SwiftUI
-                .resizable()
-                .frame(width: 100, height: 70)
-                .cornerRadius(5)
+            if let post = latestPost, let imageUrl = post.postImageUrl, !imageUrl.isEmpty {
+                let url = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0].appendingPathComponent(imageUrl)
+                if let data = try? Data(contentsOf: url), let uiImage = UIImage(data: data) {
+                    Image(uiImage: uiImage)
+                        .resizable()
+                        .aspectRatio(contentMode: .fill)
+                        .frame(width: 100, height: 70)
+                        .clipped()
+                        .cornerRadius(5)
+                } else {
+                    placeholderView
+                }
+            } else if let post = latestPost, let memo = post.memo, !memo.isEmpty {
+                Text(memo)
+                    .font(.system(size: 12, weight: .medium))
+                    .foregroundColor(.black)
+                    .frame(width: 100, height: 70)
+                    .background(Color.gray.opacity(0.3))
+                    .cornerRadius(5)
+            } else {
+                placeholderView
+            }
```
이미지나 메모 중 하나는 있는 것이 post가 생성되는 조건인데, 둘다 없을때 placeholderview가 표시되도록 예외처리 하신 게 의도된걸까요?

Gojaehyeon replied:
정상적인 상황에서는 Post 생성시 이미지, 메모 중 하나가 필수이지만, 데이터가 잘못 저장되거나 미래에 Post 생성로직이 바뀌는 등의 상황 등에서 크래시가 날 수 있기에 예외처리하였습니다.

Gojaehyeon replied:
처음 작업할때는 조건이 충족되지 않아도 작성완료가 활성화 되어있었던 터라.
현재 상황에서는 사용되지 않고 있습니다!

chawj1234 replied:
넵 감사해요!

chawj1234 commented on C3-4T2D/View/MainView/MainView.swift:
```diff
@@ -32,6 +32,7 @@ struct MainView: View {
 
     var allPostForGrid: [Post] {
         sortedProjects.flatMap { $0.postList }
+            .sorted(by: { $0.createdAt > $1.createdAt })
     }
```
Grid 뷰에서만 post가 최신순으로 정렬된 점은 의도한 부분일까요? 다른 뷰와의 정렬 기준이 달라서 혹시 특정한 이유가 있는지 궁금합니다

Gojaehyeon replied:
오우..이부분 버그가 있네요. 오래된순을 선택해도 최신순으로만 나오는데, 조치하겠습니다!

chawj1234 commented on C3-4T2D/View/Local/ProjectSectionCard.swift:
```diff
@@ -64,7 +85,7 @@ struct ProjectSectionCard: View {
+                                .font(.system(size: 19, weight: .medium))
                                 .foregroundColor(.gray)
                         }
                     }
```
피그마 Hifi 상에서는 전체 post를 모두 보여주는 것으로 이해했었는데, 현재 구현에서는 post가 4개를 초과할 경우 일부를 생략해서 보여주는 방식으로 처리된 이유가 있을까요? 변경된 의도가 있다면 공유 부탁드립니다!

Gojaehyeon replied:
4개 초과 관련 코드는 제가 짠게 아니라 확인이 필요할 듯 합니다! + 폰트만 키운거라서요!

chawj1234 reviewed (changes requested):
고생하셨습니다!

Gojaehyeon committed:
refactor: 새 프로젝트 추가시 헬퍼 함수 사용

Gojaehyeon committed:
refactor: 프로젝트 날짜 포맷팅 유틸리티 통합 및 중복 제거

- ProjectSelector에서 직접 구현한 dateString 함수 제거
- DateFormatter.projectDateRange(startDate:endDate:) 유틸리티 함수로 날짜 표기 일원화
- 코드 중복 제거 및 유지보수성 향상

Gojaehyeon committed:
refactor(model): Post의 project 관계를 옵셔널로 변경하여 cascade 삭제 지원

- Post 모델의 project 속성을 옵셔널로 변경
- Project 삭제 시 연결된 Post의 참조를 안전하게 해제할 수 있도록 개선
- SwiftData cascade delete가 정상 동작하도록 구조 변경
- 포스트가 없는 프로젝트는 바로 삭제(확인 한 번)
- 포스트가 하나라도 있으면 추가 알림(포스트 개수 안내) 후, '네, 모두 삭제' 시 포스트 → 프로젝트 순서로 안전 삭제
- 사용자 경험 및 데이터 무결성 강화

Gojaehyeon committed:
fix: 그리드뷰 포스트 정렬 버그 수정

- 그리드뷰(allPostForGrid)도 sortOrder(최신순/오래된순) - - 선택에 따라 정렬되도록 수정
- 모든 뷰에서 정렬 기준이 일관되게 동작하도록 개선

Gojaehyeon committed:
refactor: ProjectSectionCard에서 '+N' 초과 포스트 표시 기능 제거

- 프로젝트 썸네일 카드에서 4개까지만 보여주고 초과분을 '+N'으로 표시하던 기능 삭제
- 이제 프로젝트에 포함된 모든 포스트를 제한 없이 모두 썸네일로 표시
- 사용자 경험 단순화 및 정보 접근성 향상

Gojaehyeon committed:
feat: CreateView 진입시 진행중인 프로젝트에 자동 할당되는 기능 추가

chawj1234 reviewed (approved):
LGGGGTM
